closed/adds/jdk/src/share/native/jdk/crypto/jniprovider/NativeCrypto.c

```diff
@@ -44,17 +141,192 @@ typedef struct OpenSSLMDContext {
 static void printErrors(void) {
     unsigned long errCode = 0;
 
-    fprintf(stderr, "An error occurred\n");
-    while(0 != (errCode = ERR_get_error()))
+    fprintf(stderr, "An OpenSSL error occurred\n");
+    while(0 != (errCode = (*OSSL_get_error)()))
     {
         char err_str[120];
-        ERR_error_string_n(errCode, err_str, (sizeof(err_str) / sizeof(char)));
-        fprintf(stderr, "Generating error message\n" );
+        (*OSSL_error_string_n)(errCode, err_str, (sizeof(err_str) / sizeof(char)));
         fprintf(stderr, "%s\n", err_str);
     }
     fflush(stderr);
 }
 
+/*
+ * Class:     jdk_crypto_jniprovider_NativeCrypto
+ * Method:    loadCrypto
+ * Signature: ()V
+ */
+JNIEXPORT jint JNICALL Java_jdk_crypto_jniprovider_NativeCrypto_loadCrypto
+  (JNIEnv *env, jclass thisObj){
+
+    void *handle = NULL;
+    typedef const char* OSSL_version_t(int);
+
+     // Determine the version of OpenSSL.
+    OSSL_version_t* OSSL_version;
+    const char * openssl_version;
+    int ossl_ver;
+
+    // Load OpenSSL Crypto library
+    handle = load_crypto_library();
+    if (NULL == handle) {
+        //fprintf(stderr, " :FAILED TO LOAD OPENSSL CRYPTO LIBRARY\n");
+        //fflush(stderr);
+        return -1;
+    }
+
+    // Different symbols are used by OpenSSL with 1.0 and 1.1.
+    // The symbol 'OpenSSL_version' is used by OpenSSL 1.1 where as
+    // the symbol "SSLeay_version" is used by OpenSSL 1.0.
+    // Currently only openssl 1.0.2 and 1.1.0 and 1.1.1 are supported.
+    OSSL_version = (OSSL_version_t*)find_crypto_symbol(handle, "OpenSSL_version");
+
+    if (NULL == OSSL_version)  {
+        OSSL_version = (OSSL_version_t*)find_crypto_symbol(handle, "SSLeay_version");
+
+        if (NULL == OSSL_version)  {
+            //fprintf(stderr, "Only openssl 1.0.2 and 1.1.0 and 1.1.1 are supported\n");
+            //fflush(stderr);
+            unload_crypto_library(handle);
+            return -1;
+        } else {
+            openssl_version = (*OSSL_version)(0); //get OPENSSL_VERSION
+            //Ensure the OpenSSL version is "OpenSSL 1.0.2"
+            if (0 != strncmp(openssl_version, "OpenSSL 1.0.2", 13)) {
+                //fprintf(stderr, "Incompatable OpenSSL version: %s\n", openssl_version);
+                //fflush(stderr);
+                unload_crypto_library(handle);
+                return -1;
+            }
+            ossl_ver = 0;
+        }
+    } else {
+        openssl_version = (*OSSL_version)(0); //get OPENSSL_VERSION
+        //Ensure the OpenSSL version is "OpenSSL 1.1.0" or "OpenSSL 1.1.0".
+        if ((0 != strncmp(openssl_version, "OpenSSL 1.1.0", 13)) &&
+            (0 != strncmp(openssl_version, "OpenSSL 1.1.1", 13))) {
```

Comment:
This test is more restrictive than the one in `get_openssl_source.sh`: they should be the same.
Unless you can think of a reason to disallow `1.1.2`, for example, I suggest it be relaxed.

Reply:
Ok, I've relaxed it. Although looking at the version scheme OpenSSL are using (https://www.openssl.org/policies/releasestrat.html) we are unlikely to see 1.1.2 and the next version we will see is 3.0.0: Which means we'll be revisiting this all again when that happens.